backend/src/dao/ShoutoutsDao.ts

```diff
@@ -1,86 +1,59 @@
+    giver: await getMemberFromDocumentReference(dbShoutout.giver)
+  };
+}
+
+export default class ShoutoutsDao extends BaseDao<Shoutout, DBShoutout> {
+  constructor() {
+    super(shoutoutCollection, materializeShoutout, serializeShoutout);
   }
 
-  static async getShoutouts(email: string, type: 'given' | 'received'): Promise<Shoutout[]> {
+  async getAllShoutouts(): Promise<Shoutout[]> {
+    return this.getAllDocuments();
+  }
+
+  async getShoutouts(email: string, type: 'given' | 'received'): Promise<Shoutout[]> {
     const givenOrReceived = type === 'given' ? 'giver' : 'receiver';
-    const shoutoutRefs = await shoutoutCollection
+    const shoutoutRefs = await this.collection
       .where(givenOrReceived, '==', memberCollection.doc(email))
       .get();
     return Promise.all(
-      shoutoutRefs.docs.map(async (shoutoutRef) => {
-        const { giver, receiver, message, isAnon, timestamp, hidden, uuid } = shoutoutRef.data();
-        return {
-          giver: await getMemberFromDocumentReference(giver),
-          receiver,
-          message,
-          isAnon,
-          timestamp,
-          hidden,
-          uuid
-        };
-      })
+      shoutoutRefs.docs.map(async (shoutoutRef) => materializeShoutout(shoutoutRef.data()))
     );
   }
 
-  static async getShoutout(uuid: string): Promise<Shoutout | undefined> {
-    const doc = await shoutoutCollection.doc(uuid).get();
-    const dbShoutout = doc.data();
-    if (!dbShoutout) return undefined;
-    return {
-      ...dbShoutout,
-      giver: await getMemberFromDocumentReference(dbShoutout.giver)
-    };
+  async getShoutout(uuid: string): Promise<Shoutout | null> {
+    return this.getDocument(uuid);
   }
 
-  static async setShoutout(shoutout: {
-    giver: IdolMember;
-    receiver: string;
-    message: string;
-    isAnon: boolean;
-    timestamp: number;
-    hidden: boolean;
-    uuid: string;
-  }): Promise<Shoutout> {
-    const shoutoutRef: DBShoutout = {
+  async createShoutout(shoutout: Shoutout): Promise<Shoutout> {
+    const shoutoutWithUUID = {
       ...shoutout,
-      giver: memberCollection.doc(shoutout.giver.email),
       uuid: shoutout.uuid ? shoutout.uuid : uuidv4()
     };
-    await shoutoutCollection.doc(shoutoutRef.uuid).set(shoutoutRef);
-    return shoutout;
+    return this.createDocument(shoutoutWithUUID.uuid, shoutoutWithUUID);
   }
 
-  static async updateShoutout(shoutout: Shoutout): Promise<Shoutout> {
-    const shoutoutDoc = shoutoutCollection.doc(shoutout.uuid);
-    const ref = await shoutoutDoc.get();
-    if (!ref.exists) throw new NotFoundError(`No shoutout '${shoutout.uuid}' exists.`);
-    const shoutoutRef: DBShoutout = {
-      ...shoutout,
-      giver: memberCollection.doc(shoutout.giver.email)
-    };
-    await shoutoutCollection.doc(shoutout.uuid).update(shoutoutRef);
-    return shoutout;
+  async updateShoutout(shoutout: Shoutout): Promise<Shoutout> {
+    return this.updateDocument(shoutout.uuid, shoutout);
```

Comment:
Along with the concrete and db types we parameterize the base dao on, could we pass information about which field should be used as the doc id (email for members, uuid for most other things)? Or even just an assumption that it will be `uuid`, and we can override it in case of exceptions.

Reply:
This would also allow us to implement the common `uuid ? uuid : uuidv4()` check by default, which could remove forgetting this as a potential source for bugs in the future

Reply:
I think it's hard to be able to make modifications to the data in `BaseDao` since we're working with generic types. TypeScript doesn't know that the type `E` or `D` has a field `uuid`. If we really wanted to make it so that we assume that there is a `uuid` field or to implement the common `uuid ? uuid : uuidv4()` logic, we'd have to use an interface like 
```
interface ObjectWithUUID {
  uuid: string;
}
```
that all of the types with uuid implement and we use this type instead of the generic types in `BaseDao`. But then the issue with this is that we can't use `BaseDao` anymore for members or sign-in since neither of those types have a uuid field. 

Similarly we also can't really pass information about which field should be used as the doc id since TypeScript doesn't know that that type exists on some generic type.